### ISSUE TITLE
phpunit.xsd not extracted on composer install

### ISSUE THREAD

**Issue by phil-davis** [sebastianbergmann/phpunit#3572]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 7.5.7
| PHP version         | 7.2.16
| Installation Method | Composer

When I run with `phpunit` 7 I get this kind of output:
```
./lib/composer/phpunit/phpunit/phpunit -c tests/phpunit-autotest.xml tests/lib/DB/ConnectionTest.php 
PHP Warning:  DOMDocument::schemaValidate(): Invalid Schema in /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/Util/Configuration.php on line 1034
PHP Stack trace:
PHP   1. {main}() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit\TextUI\Command::main() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/phpunit:61
PHP   3. PHPUnit\TextUI\Command->run() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/TextUI/Command.php:162
PHP   4. PHPUnit\TextUI\Command->handleArguments() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/TextUI/Command.php:173
PHP   5. PHPUnit\Util\Configuration::getInstance() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/TextUI/Command.php:836
PHP   6. PHPUnit\Util\Configuration->__construct() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/Util/Configuration.php:206
PHP   7. PHPUnit\Util\Configuration->validateConfigurationAgainstSchema() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/Util/Configuration.php:223
PHP   8. DOMDocument->schemaValidate() /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/src/Util/Configuration.php:1034
PHPUnit 7.5.7 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.16-1+ubuntu18.04.1+deb.sury.org+1 with Xdebug 2.7.0
Configuration: /home/phil/git/owncloud/core/tests/phpunit-autotest.xml

  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 0:
  - failed to load external entity "/home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/phpunit.xsd"
  - Failed to locate the main schema resource at '/home/phil/git/owncloud/core/lib/composer/phpunit/phpunit/phpunit.xsd'.

  Test results may not be as expected.


.......                                                             7 / 7 (100%)

Time: 1.06 seconds, Memory: 18.00 MB

OK (7 tests, 10 assertions)
```
Tests still work, but the big warning does not feel good ;)

I cleaned my composer cache, and made sure to install the latest composer `1.8.4`
After `composer install` I end up with a cache of the `phpunit` that it downloaded:
```
ls -l .cache/composer/files/phpunit/phpunit/
total 764
-rw-r--r-- 1 phil phil 779332 Mar 22 19:08 1bac195c7794553c5ffffd4542a000fcc9d85184.zip
```

In that zip file there is a `phpunit.xsd` - good. But it has not been unpacked:
```
ls -l /home/phil/git/owncloud/core/lib/composer/phpunit/phpunit
total 16
-rw-r--r-- 1 phil phil 2334 Mar 16 13:16 composer.json
-rw-r--r-- 1 phil phil 1542 Mar 16 13:16 LICENSE
-rwxr-xr-x 1 phil phil 1408 Mar 16 13:16 phpunit
drwxr-xr-x 6 phil phil 4096 Mar 16 13:16 src
```

PR #3085 "Add phpunit.xsd into the phar file" made a change to make sure that `phpunit.xsd` got into the `phar`.

Is there something that needs to be done to convince `composer` that it should extract `phpunit.xsd` when doing an install?

Output of `composer info | sort` as requested:
```
bamarni/composer-bin-plugin           v1.2.0            
bantu/ini-get-wrapper                 v1.0.1             Convenience wrapper around ini_get()
composer/semver                       1.5.0              Semver library that offers utilities, version constraint parsing and validation.
container-interop/container-interop   1.2.0              Promoting the interoperability of container objects (DIC, SL, etc.)
deepdiver1975/tarstreamer             v0.1.0             A library for dynamically streaming dynamic tar files without the need to have the complete file stored on...
deepdiver/zipstreamer                 1.1.1              Stream zip files without i/o overhead
dg/composer-cleaner                   v2.0               Victor The Cleaner: removes unnecessary files from vendor directory.
doctrine/cache                        v1.8.0             Caching library offering an object-oriented API for many cache backends
doctrine/dbal                         v2.9.2             Powerful PHP database abstraction layer (DBAL) with many features for database schema introspection and ma...
doctrine/event-manager                v1.0.0             Doctrine Event Manager component
doctrine/instantiator                 1.2.0              A small, lightweight utility to instantiate objects in PHP without invoking their constructors
doctrine/lexer                        v1.0.1             Base library for a lexer that can be used in Top-Down, Recursive Descent Parsers.
egulias/email-validator               2.1.7              A library for validating emails against several RFCs
guzzle/common                         v3.8.1             Common libraries used by Guzzle
guzzlehttp/guzzle                     5.3.3              Guzzle is a PHP HTTP client library and framework for building RESTful web service clients
guzzlehttp/ringphp                    1.1.1              Provides a simple API and specification that abstracts away the details of HTTP into a single PHP function.
guzzlehttp/streams                    3.0.0              Provides a simple abstraction over streams of data
guzzle/http                           v3.8.1             HTTP libraries used by Guzzle
guzzle/parser                         v3.8.1             Interchangeable parsers used by Guzzle
guzzle/stream                         v3.8.1             Guzzle stream wrapper component
icewind/streams                       0.5.2              A set of generic stream wrappers
interfasys/lognormalizer              v1.0               Parses variables and converts them to string so that they can be logged
jakub-onderka/php-console-color       v0.2              
jakub-onderka/php-console-highlighter v0.4               Highlight PHP code in terminal
jeremeamia/SuperClosure               2.4.0              Serialize Closure objects, including their context and binding
league/flysystem                      1.0.50             Filesystem abstraction: Many filesystems, one API.
lukasreschke/id3parser                v0.0.3            
mikey179/vfsStream                    v1.6.5             Virtual file system to mock the real file system in unit tests.
myclabs/deep-copy                     1.8.1              Create deep copies (clones) of your objects
nikic/php-parser                      v4.2.1             A PHP parser written in PHP
patchwork/jsqueeze                    v2.0.5             Efficient JavaScript minification in PHP
patchwork/utf8                        v1.3.1             Portable and performant UTF-8, Unicode and Grapheme Clusters for PHP
pear/archive_tar                      1.4.6              Tar file management class with compression support (gzip, bzip2, lzma2)
pear/console_getopt                   v1.4.2             More info available on: http://pear.php.net/package/Console_Getopt
pear/pear-core-minimal                v1.10.9            Minimal set of PEAR core files to be used as composer dependency
pear/pear_exception                   v1.0.0             The PEAR Exception base class.
phar-io/manifest                      1.0.3              Component for reading phar.io manifest information from a PHP Archive (PHAR)
phar-io/version                       2.0.1              Library for handling version information and constraints
phpdocumentor/reflection-common       1.0.1              Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock     4.3.0              With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve inf...
phpdocumentor/type-resolver           0.4.0             
phpseclib/phpseclib                   2.0.15             PHP Secure Communications Library - Pure-PHP implementations of RSA, AES, SSH2, SFTP, X.509 etc.
phpspec/prophecy                      1.8.0              Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage             6.1.4              Library that provides collection, processing, and rendering functionality for PHP code coverage information.
phpunit/php-file-iterator             2.0.2              FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-text-template             1.2.1              Simple template engine.
phpunit/php-timer                     2.1.1              Utility class for timing
phpunit/php-token-stream              3.0.1              Wrapper around PHP's tokenizer extension.
phpunit/phpunit                       7.5.7              The PHP Unit Testing framework.
pimple/pimple                         v3.2.3             Pimple, a simple Dependency Injection Container
psr/container                         1.0.0              Common Container Interface (PHP FIG PSR-11)
psr/log                               1.1.0              Common interface for logging libraries
punic/punic                           3.3.1              PHP-Unicode CLDR
rackspace/php-opencloud               v1.9.2             PHP SDK for Rackspace/OpenStack APIs
react/promise                         v2.7.1             A lightweight implementation of CommonJS Promises/A for PHP
roave/security-advisories             dev-master 018ec51 Prevents installation of composer packages with known security vulnerabilities: no API, simply require it
sabre/dav                             4.0.0-alpha5       WebDAV Framework for PHP
sabre/event                           5.0.3              sabre/event is a library for lightweight event-based programming
sabre/http                            5.0.0              The sabre/http library provides utilities for dealing with http requests and responses. 
sabre/uri                             2.1.1              Functions for making sense out of URIs.
sabre/vobject                         4.2.0              The VObject library for PHP allows you to easily parse and manipulate iCalendar and vCard objects
sabre/xml                             2.1.2              sabre/xml is an XML library that you may not hate.
sebastian/code-unit-reverse-lookup    1.0.1              Looks up which function or method a line of code belongs to
sebastian/comparator                  3.0.2              Provides the functionality to compare PHP values for equality
sebastian/diff                        3.0.2              Diff implementation
sebastian/environment                 4.1.0              Provides functionality to handle HHVM/PHP environments
sebastian/exporter                    3.1.0              Provides the functionality to export PHP variables for visualization
sebastian/global-state                2.0.0              Snapshotting of global state
sebastian/object-enumerator           3.0.3              Traverses array structures and object graphs to enumerate all referenced objects
sebastian/object-reflector            1.1.1              Allows reflection of object attributes, including inherited and non-public ones
sebastian/recursion-context           3.0.0              Provides functionality to recursively process PHP variables
sebastian/resource-operations         2.0.1              Provides a list of PHP built-in functions that operate on resources
sebastian/version                     2.0.1              Library that helps with managing the version number of Git-hosted PHP projects
swiftmailer/swiftmailer               v6.2.0             Swiftmailer, free feature-rich PHP mailer
symfony/console                       v3.4.23            Symfony Console Component
symfony/debug                         v3.4.23            Symfony Debug Component
symfony/event-dispatcher              v3.4.23            Symfony EventDispatcher Component
symfony/polyfill-ctype                v1.10.0            Symfony polyfill for ctype functions
symfony/polyfill-iconv                v1.10.0            Symfony polyfill for the Iconv extension
symfony/polyfill-intl-idn             v1.10.0            Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-mbstring             v1.10.0            Symfony polyfill for the Mbstring extension
symfony/polyfill-php56                v1.10.0            Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php72                v1.10.0            Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-util                 v1.10.0            Symfony utilities for portability of PHP codes
symfony/process                       v3.4.23            Symfony Process Component
symfony/routing                       v3.4.23            Symfony Routing Component
symfony/translation                   v3.4.23            Symfony Translation Component
theseer/tokenizer                     1.1.0              A small library for converting tokenized PHP source code into XML and potentially other formats
webmozart/assert                      1.4.0              Assertions to validate method input/output with nice error messages.
zendframework/zend-filter             2.9.1              provides a set of commonly needed data filters
zendframework/zend-inputfilter        2.10.0             Normalize and validate input sets from the web, APIs, the CLI, and more, including files
zendframework/zend-servicemanager     3.4.0              Factory-Driven Dependency Injection Container
zendframework/zend-stdlib             3.2.1              SPL extensions, array utilities, error handlers, and more
zendframework/zend-validator          2.12.0             provides a set of commonly needed validators
```

**Comment by phil-davis** [comment]
Note: if I manually extract `phpunit.xsd` into `/home/phil/git/owncloud/core/lib/composer/phpunit/phpunit` then the test warning goes away.

**Comment by sebastianbergmann** [comment]
I cannot reproduce this:
```
$ composer require --dev phpunit/phpunit
Using version ^8.0 for phpunit/phpunit
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 28 installs, 0 updates, 0 removals
  - Installing sebastian/version (2.0.1): Loading from cache
  - Installing sebastian/resource-operations (2.0.1): Loading from cache
  - Installing sebastian/recursion-context (3.0.0): Loading from cache
  - Installing sebastian/object-reflector (1.1.1): Loading from cache
  - Installing sebastian/object-enumerator (3.0.3): Loading from cache
  - Installing sebastian/global-state (3.0.0): Loading from cache
  - Installing sebastian/exporter (3.1.0): Loading from cache
  - Installing sebastian/environment (4.1.0): Loading from cache
  - Installing sebastian/diff (3.0.2): Loading from cache
  - Installing sebastian/comparator (3.0.2): Loading from cache
  - Installing phpunit/php-timer (2.1.1): Loading from cache
  - Installing phpunit/php-text-template (1.2.1): Loading from cache
  - Installing phpunit/php-file-iterator (2.0.2): Loading from cache
  - Installing theseer/tokenizer (1.1.0): Loading from cache
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Loading from cache
  - Installing phpunit/php-token-stream (3.0.1): Loading from cache
  - Installing phpunit/php-code-coverage (7.0.3): Loading from cache
  - Installing doctrine/instantiator (1.2.0): Loading from cache
  - Installing symfony/polyfill-ctype (v1.11.0): Loading from cache
  - Installing webmozart/assert (1.4.0): Loading from cache
  - Installing phpdocumentor/reflection-common (1.0.1): Loading from cache
  - Installing phpdocumentor/type-resolver (0.4.0): Loading from cache
  - Installing phpdocumentor/reflection-docblock (4.3.0): Loading from cache
  - Installing phpspec/prophecy (1.8.0): Loading from cache
  - Installing phar-io/version (2.0.1): Loading from cache
  - Installing phar-io/manifest (1.0.3): Loading from cache
  - Installing myclabs/deep-copy (1.8.1): Loading from cache
  - Installing phpunit/phpunit (8.0.5): Loading from cache
sebastian/global-state suggests installing ext-uopz (*)
phpunit/php-code-coverage suggests installing ext-xdebug (^2.6.1)
phpunit/phpunit suggests installing phpunit/php-invoker (^2.0)
phpunit/phpunit suggests installing ext-xdebug (*)
Writing lock file
Generating autoload files
```
```
$ ll vendor/phpunit/phpunit/phpunit.xsd 
-rw-r--r--. 1 sb sb 17K Mar 16 08:33 vendor/phpunit/phpunit/phpunit.xsd
```

**Comment by sebastianbergmann** [comment]
@Seldaek @naderman Do you have an idea what might be going on here? Thanks!

**Comment by naderman** [comment]
The fact the phpunit ends up being installed into "/lib/composer/phpunit/phpunit" suggests that you are using some kind of Composer plugin which changes composer install behavior, since composer otherwise always uses a directory called "vendor". The problem is most likely with this plugin. Please post your full composer.json, or better yet, if you can figure out what's installing phpunit into this special location, get in touch with whoever maintains that plugin.

**Comment by sebastianbergmann** [comment]
Thank you, @naderman.

**Comment by phil-davis** [comment]
`owncloud/core` is my "example". https://github.com/owncloud/core/blob/master/composer.json
Reproduced here:
```
{
    "name": "owncloud/core",
    "description": "A safe home for all your data",
    "license": "AGPL-1.0",
    "config" : {
        "vendor-dir": "lib/composer",
        "optimize-autoloader": true,
        "classmap-authoritative": false,
        "platform": {
            "php": "7.1"
        }
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "autoload" : {
        "psr-4": {
            "OC\\": "lib/private",
            "OC\\Core\\": "core/",
            "OC\\Settings\\": "settings/",
            "OCP\\": "lib/public"
        },
        "classmap": ["lib/private/legacy"]
    },
    "autoload-dev" : {
        "files": ["tests/lib/TestCase.php"]
    },
    "require-dev": {
        "bamarni/composer-bin-plugin": "^1.2",
        "jakub-onderka/php-console-highlighter": "^0.4",
        "mikey179/vfsstream": "^1.6",
        "phpunit/phpunit": "^7.5",
        "roave/security-advisories": "dev-master"
    },
    "require": {
        "php": ">=7.1",
        "doctrine/dbal": "^2.5",
        "phpseclib/phpseclib": "^2.0",
        "rackspace/php-opencloud": "v1.9.2",
        "jeremeamia/superclosure": "^2.4",
        "bantu/ini-get-wrapper": "v1.0.1",
        "punic/punic": "^3.1",
        "pear/archive_tar": "1.4.6",
        "patchwork/utf8": "^1.3",
        "symfony/console": "^3.1",
        "symfony/event-dispatcher": "^3.1",
        "symfony/routing": "^3.1",
        "symfony/process": "^3.1",
        "pimple/pimple": "3.2.3",
        "nikic/php-parser": "^4.0",
        "icewind/streams": "0.5.2",
        "swiftmailer/swiftmailer": "^6.0",
        "guzzlehttp/guzzle": "^5.3",
        "league/flysystem": "^1.0",
        "pear/pear-core-minimal": "^v1.10",
        "interfasys/lognormalizer": "^v1.0",
        "deepdiver1975/tarstreamer": "v0.1.0",
        "patchwork/jsqueeze": "^2.0",
        "lukasreschke/id3parser": "^0.0.3",
        "sabre/dav": "^4.0.0-alpha5",
        "deepdiver/zipstreamer": "^1.1",
        "symfony/translation": "^3.4",
        "zendframework/zend-inputfilter": "^2.8",
        "zendframework/zend-servicemanager": "^3.3",
        "zendframework/zend-validator": "^2.10",
        "composer/semver": "^1.4",
        "ext-json": "*",
        "sabre/vobject": "^4.2",
        "dg/composer-cleaner": "^2.0"
    },
    "extra": {
        "bamarni-bin": {
            "bin-links": false
        }
    }
}
```

I tries removing `"vendor-dir": "lib/composer"` and I have the same trouble in, now in the `vendor` dir.

I will play with other dependencies - `bamarni/composer-bin-plugin` looks like the first suspect to remove :)

**Comment by phil-davis** [comment]
The culprit was `dg/composer-cleaner` - it is way too enthusiastic about deleting stuff it thinks is not useful! I should have noticed that a long time earlier. I saw the PR here about `phpunit.xsd` and got side-tracked thinking my problem might have been somehow related.

I had to add the `cleaner-ignore` section below:
```
    "extra": {
        "bamarni-bin": {
            "bin-links": false
        },
        "cleaner-ignore": {
            "phpunit/phpunit": ["phpunit.xsd"]
        }
    }
```

Thanks for the comments and sorry for the noise. Hopefully at least this thread might point someone else in the right direction in the future.
